pkg/nsp/registry/registry.go

```diff
@@ -0,0 +1,130 @@
+/*
+Copyright (c) 2021 Nordix Foundation
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package registry
+
+import (
+	nspAPI "github.com/nordix/meridio/api/nsp/v1"
+	"github.com/nordix/meridio/pkg/nsp/types"
+)
+
+type targetKey struct {
+	Ips    []string
+	stream *nspAPI.Stream
+}
+
+type TargetRegistryMemory struct {
+	targets []*nspAPI.Target
+	Chan    chan<- struct{}
+}
+
+func New(c chan<- struct{}) types.TargetRegistry {
+	targetRegistryMemory := &TargetRegistryMemory{
+		Chan:    c,
+		targets: []*nspAPI.Target{},
+	}
+	return targetRegistryMemory
+}
+
+func (trm *TargetRegistryMemory) Set(target *nspAPI.Target) {
+	index := trm.getIndex(target)
+	if index == -1 { // add
+		trm.targets = append(trm.targets, target)
+		trm.notify()
+		return
+	}
+	//update
+	trm.targets[index].Context = target.GetContext()
+	trm.targets[index].Status = target.GetStatus()
+	trm.targets[index].Type = target.GetType()
```

Comment:
I'm not sure it should be allowed for Type to be changed. Do you have any particular reason for adding support for that?
IMHO Type should be immutable, and there should be no interference among Targets with different Types. Meaning that IPs should be made re-usable without interference across the different Target types. (I.e. Type should be made part of the key.)

I guess the upgradeability of the Type was the driving force behind using a common storage for the different target Types as well. However if I got it right, this target nsp service is planned to be a generic service that could be used by many other target types in the future. If so, the common shared array for the different targets could have a performance penalty. The current approach should be fine for now though.